### DISCUSSION TITLE
set_global_routing_layer_adjustment

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#5700] [General]
So set_global_routing_layer_adjustment is essentially setting some global variables...

Is there a reason why global variables are used here instead of arguments to global_route?

I much prefer arguments to global route than a global variable, if those are the choices. It is easier to understand the ORFS .tcl scripts when I can see what information flows where...

Same applies to set_macro_extension.

For instance, I wonder why gui.tcl sets these global global route variables...

**Comment by maliberty** [comment]
The adjustments are needed by Rudy (global_place -routability & estimated congestion heatmap) as well.

**Reply by oharboe** [reply]
I see.

This underscores my point: it would be more clear with arguments than global variables.

**Reply by maliberty** [reply]
There are no arguments to heatmaps

**Reply by oharboe** [reply]
I see...

Hard to follow, a bit of surgery here and there....

Could this be recorded in .odb?

**Reply by maliberty** [reply]
Sure (though that is a global of another form).

**Reply by oharboe** [reply]
indeed, like everything in .odb....

**Reply by oharboe** [reply]
In bazel-orfs and also for documentation purposes, it would be nice to have a machine verified/readable mapping from environment variable name to which stages it is used in...

Along the lines of:

https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/2331

**Reply by oharboe** [reply]
@jeffng-or FYI

**Reply by oharboe** [reply]
@jeffng-or In MegaBoom, I have to be [weary](https://github.com/The-OpenROAD-Project/megaboom/pull/106) that global route arguments are being used in placement... 

The alternative, rebuilding everything from scratch isn't great.

Passing all variables to all stages, forces me to rebuild everything always, which would mean in this case, having to wait 5 hours, before global placement even starts...

If ORFS could give me a machine readable description of exactly which variables are used where, bazel-orfs could use this information to handle all of this automatically.

**Reply by maliberty** [reply]
I see no way to automatically generate that so it would just be moving the manual work around and subject the same human error.

**Reply by oharboe** [reply]
Except it has to be done correctly once by someone who has a chance of finding out(ORFS maintainers), instead of every time a user uses this information.

I did add some automation to try to force errors if a variable is used outside of its defined scope, so it is not entirely "subject to the same human error".

Other suggestions?

**Reply by maliberty** [reply]
No users of ORFS need this information - its purely a bazel requirement.  If you have a check that seems sufficient.

**Reply by oharboe** [reply]
With the caveat that anyone using global placement needs to know that the global routing arguments are required for global placement to work correctly since it now relies on RUDY, right?

**Reply by oharboe** [reply]
Non-ORFS users that I know of are OpenLane, SiliconCompiler,  bazel_rules_hdl and bazel-orfs.

**Reply by maliberty** [reply]
Also hammer and doubtless more than neither of us knows about.   Developers of flows take this burden on themselves (and quite a lot more).

**Comment by oharboe** [comment]
Here bazel-orfs has been modified to map variables onto stages using a hardcoded mapping from arguments to stages: https://github.com/The-OpenROAD-Project/bazel-orfs/pull/140
